### PR TITLE
Improve TERM color detection support

### PR DESCRIPTION
**Pull request by nevans**
Adds better color support detection based on TERM variable.  Ultimately,
it would probably be better if we could delegate to the terminfo DB for
this.  But relying on "color" or "256color" or "direct" is fairly safe.

### CODE REVIEW

File: lib/test/unit/color-scheme.rb [hunk 1]
```diff
@@ -5,9 +5,23 @@ module Unit
     class ColorScheme
       include Enumerable
 
+      TERM_256 = /
+        [+-]256color|
+        ^(alacritty|
```
Comment (kou) [inline lib/test/unit/color-scheme.rb:10]: Could you use `\A` instead of `^`?

File: lib/test/unit/color-scheme.rb [hunk 2]
```diff
@@ -5,9 +5,23 @@ module Unit
...
+          iTerm\s?\d*\.app|
+          kitty|
+          mintty|
+          ms-terminal|
+          nsterm|
+          nsterm-build\d+|
+          terminator|
+          terminology(-[0-9.]+)?|
+          termite|
+          vscode)$/x
```
Comment (kou) [inline lib/test/unit/color-scheme.rb:20]: Could you use `\z` instead of `$`?

Comment (kou) [reply lib/test/unit/color-scheme.rb:20]: Could you sort this list in alphabetical order?

File: lib/test/unit/color-scheme.rb [hunk 3]
```diff
@@ -140,7 +154,9 @@ def guess_available_colors_from_colorterm_env
 
         def guess_available_colors_from_term_env
           case ENV["TERM"]
-          when /-256color\z/
+          when /[+-]direct/
```
Comment (kou) [inline lib/test/unit/color-scheme.rb:157]: `xterm-direct16` matches this. Is it expected?

File: lib/test/unit/ui/console/testrunner.rb [hunk 4]
```diff
@@ -540,13 +540,27 @@ def summary_marker_color
...
           end
 
+          TERM_COLOR_SUPPORT = /
+            color|  # explicitly claims color support in the name
+            direct| # explicitly claims "direct color" (24 bit) support
+            #{ColorScheme::TERM_256}
+            ^cygwin|
+            ^linux|
+            ^nsterm-bce|
+            ^nsterm-c-.*|
```
Comment (kou) [inline lib/test/unit/ui/console/testrunner.rb:550]: `.*` is needless.

File: test/test-color-scheme.rb [hunk 5]
```diff
@@ -45,12 +45,14 @@ class TestFor8Colors < self
     def setup
       @original_term, ENV["TERM"] = ENV["TERM"], nil
       @original_color_term, ENV["COLORTERM"] = ENV["COLORTERM"], nil
+      @original_vte_version, ENV["VTE_VERSION"] = ENV["VTE_VERSION"], nil
```
Comment (kou) [inline test/test-color-scheme.rb:48]: Why is this needed?

File: test/test-color-scheme.rb [hunk 6]
```diff
@@ -78,5 +80,35 @@ def test_default
       assert_equal(expected_schema_keys.sort,
                    Test::Unit::ColorScheme.default.to_hash.keys.sort)
     end
+
+    def test_available_colors
```
Comment (kou) [inline test/test-color-scheme.rb:84]: `TestFor8Colors` is a test case for 8 colors case.
Here is not suitable location. 
Newly test case such as `TestGuess` or something is better.

File: test/test-color-scheme.rb [hunk 7]
```diff
@@ -78,5 +80,35 @@ def test_default
...
+        "iTerm.app"            => 256,
+        "screen-256color"      => 256,
+        "screenxterm-256color" => 256,
+        "tmux-256color"        => 256,
+        "vte-256color"         => 256,
+        "vscode-direct"        => 2**24,
+        "vte-direct"           => 2**24,
+        "xterm-direct"         => 2**24,
+      }
+      terms.each do |term, expected_available|
```
Comment (kou) [inline test/test-color-scheme.rb:98]: We should not iterate test patterns.
We should use data driven feature for this: https://github.com/test-unit/test-unit/blob/master/test/test-data.rb#L40-L45

File: test/test-color-scheme.rb [hunk 8]
```diff
@@ -78,5 +80,35 @@ def test_default
...
+        ENV["TERM"] = term
+        assert_equal(expected_available,
+                     Test::Unit::ColorScheme.available_colors,
+                     "Incorrect available_colors for TERM=%s" % [term])
+      end
+    end
+
+    def test_default_for_direct_color
+      ENV["TERM"] = "xterm-direct"
+      assert_equal(2**24, Test::Unit::ColorScheme.available_colors)
```
Comment (kou) [inline test/test-color-scheme.rb:108]: This is needless.
This is tested in other test.

**Review by kou — CHANGES_REQUESTED** [review]

**Comment by nevans** [comment]
Thanks for all of the feedback! I'll respond soon.

File: test/test-color-scheme.rb (hunk 5 above)
Comment (nevans) [reply test/test-color-scheme.rb:48]: This is needed because otherwise `guess_available_colors_from_vte_version_env` will override the result of `available_colors` whenever `VTE_VERSION` is set.

File: lib/test/unit/color-scheme.rb (hunk 3 above)
Comment (nevans) [reply lib/test/unit/color-scheme.rb:157]: I didn't notice `xterm-direct16` (nor `xterm-direct256`) in any of my terminfo DBs, so I didn't test those.  But I just looked them up and found that they're both in the current upstream ncurses terminfo DB. I looked into it and... **the short tldr answer:** _yes_, they should all match as 2**24 colors. :slightly_smiling_face:

To understand why, parse terminfo's `use=...` inheritence and see that `xterm+direct16` and `xterm+direct256` are both defined with `use=xterm+direct` and both inherit `colors#0x1000000` even though they override `CO` (to 16 and 256, respectively).  If you see any of these `TERM` values, they should _all_ be able to handle the actual direct color escape sequences. The difference is just about how ncurses/`tput`/terminfo translate a single `int` into that escape sequence.

The much much longer answer explains why. It isn't really necessary, but a couple of weeks ago I put in the time to figure out how terminfo stack-based parameterized strings work, so I may as well share. :laughing: 

`tput` (and C functions `tparm`/`tiparm`) still only take a single integer value for `setaf`/`setab`, and unfortunately ncurses/terminfo hasn't added something like `setf_rgb`/`setb_rgb`. So they overload it: `0-7` are still the original ANSI colors, and `8-0xFFFFFF` are the "direct" colors, i.e. not indexed (and optionally initialized via `initc` or `initp`). This omits the 7 darkest shades of blue from direct addressing and any of the bright versions (8-15) of those ANSI colors, and the remaining 256 indexed colors (which default to 6x6x6 color cube + 24 shades of grey, which according to the ncurses comment author are `"of interest only to the numerically illiterate"`. :laughing:

Yes, once you have direct colors, you can translate everything to RGB. But if you have software that doesn't know about direct RGB colors and assumes it can expect the default 8-255 values to work, it won't get the intended colors but will create all of the shades of blue (without red or green).

So here are the terminfo values for `xterm-direct`, `xterm-direct16`, and `xterm-direct256`:
```
xterm+direct|xterm with direct-color indexing,
        RGB,
        colors#0x1000000, pairs#0x10000, CO#8,
        initc@, op=\E[39;49m,
        setab=\E[%?%p1%{8}%<%t4%p1%d%e48\:2\:\:%p1%{65536}%/%d\:%p1
              %{256}%/%{255}%&%d\:%p1%{255}%&%d%;m,
        setaf=\E[%?%p1%{8}%<%t3%p1%d%e38\:2\:\:%p1%{65536}%/%d\:%p1
              %{256}%/%{255}%&%d\:%p1%{255}%&%d%;m,
        setb@, setf@,
xterm-direct|xterm with direct-color indexing,
        use=xterm+direct, use=xterm,

xterm+direct16|xterm with direct-color indexing,
        CO#16,
        setab=\E[%?%p1%{8}%<%t4%p1%d%e%?%p1%{16}%<%t%p1%{92}%+%e48\:
              2\:\:%p1%{65536}%/%d\:%p1%{256}%/%{255}%&%d\:%p1%{255}
              %&%d%;%;m,
        setaf=\E[%?%p1%{8}%<%t3%p1%d%e%?%p1%{16}%<%t%p1%'R'%+%e38\:2
              \:\:%p1%{65536}%/%d\:%p1%{256}%/%{255}%&%d\:%p1%{255}
              %&%d%;%;m,
        setb@, setf@, use=xterm+direct,
xterm-direct16|xterm with direct-colors and 16 indexed colors,
        use=xterm, use=xterm+direct16,

xterm+direct256|xterm with direct-color indexing,
        CO#0x100,
        setab=\E[%?%p1%{8}%<%t4%p1%d%e%p1%{16}%<%t10%p1%{8}%-%d%e%?
              %p1%{256}%<%t48;5;%p1%d%e48\:2\:\:%p1%{65536}%/%d\:%p1
              %{256}%/%{255}%&%d\:%p1%{255}%&%d%;%;m,
        setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e%?
              %p1%{256}%<%t38;5;%p1%d%e38\:2\:\:%p1%{65536}%/%d\:%p1
              %{256}%/%{255}%&%d\:%p1%{255}%&%d%;%;m,
        setb@, setf@, use=xterm+direct,
xterm-direct256|xterm with direct-colors and 256 indexed colors,
        use=xterm, use=xterm+direct256,
```

I'll (try to accurately) translate two of those `setaf` functions into ruby, so you don't need to flip back and forth between `man 5 terminfo` and this, like I did a week or so ago :slightly_smiling_face:

```ruby
# Keep in mind the following:
#
# The strings operate a stack, so operators use postfix notation
# %? cond %t then %e else %;
#
# (0xRRGGBB / 0x10000) = 0xRR
# (0xRRGGBB / 0x100) & 0xFF = 0xGG
# (0xRRGGBB & 0xFF) = 0xBB

#        setaf=\E[%?%p1%{8}%<%t3%p1%d%e38\:2\:\:%p1%{65536}%/%d\:%p1
#              %{256}%/%{255}%&%d\:%p1%{255}%&%d%;m
def xterm_direct_setaf(p1)
  print "\e["                         # \E[
  if p1 < 8                           # %?%p1%{8}%<%t
    print "3%d" % [p1]                # 3%p1%d
  else                                # %e
    print "38:2:"                     # 38\:2\:
    print ":%d" % [p1 / 65536]        # \:%p1%{65536}%/%d
    print ":%d" % [(p1 / 256) & 255]  # \:%p1%{256}%/%{255}%&%d
    print ":%d" % [p1 & 255]          # \:%p1%{255}%&%d
  end                                 # %;
  print "m"                           # m
end

#        setaf=\E[%?%p1%{8}%<%t3%p1%d%e%p1%{16}%<%t9%p1%{8}%-%d%e%?
#              %p1%{256}%<%t38;5;%p1%d%e38\:2\:\:%p1%{65536}%/%d\:%p1
#              %{256}%/%{255}%&%d\:%p1%{255}%&%d%;%;m
def xterm_direct256_setaf(p1)
  print "\e["                           # \E[
  if p1 < 8                             # %? %p1%{8}%< %t
    print "3%d" % [p1]                  # 3%p1%d
  elsif p1 < 16                         # %p1%{16}%< %t
    print "9%d" % [p1-8]                # 9%p1%{8}%-%d
  else                                  # %e
    if p1 < 256                         # %? %p1%{256}%< %t
      print "38;5;%d" % [p1]            # 38;5;%p1%d
    else                                # %e
      print "38:2:"                     # 38\:2\:
      print ":%d" % [p1 / 65536]        # \:%p1%{65536}%/%d
      print ":%d" % [(p1 / 256) & 255]  # \:%p1%{256}%/%{255}%&%d
      print ":%d" % [p1 & 255]          # \:%p1%{255}%&%d
    end                                 # %;
  end                                   # %;
  print "m"                             # m
end
```

So `xterm+direct256` gets back the ability to have `tput 230` do what you might expect, but at the expense of losing the entire spectrum of blue where R=0, G=0. This is easy to work around by setting R=1 and G=1, but it's a bit of a hack!

IMO: ncurses/terminfo should have simply added new `set_rgb_f` and `set_rgb_b` functions and left the old ones with their 256color meaning. :)  Oh well!

File: lib/test/unit/ui/console/testrunner.rb (hunk 4 above)
Comment (nevans) [reply lib/test/unit/ui/console/testrunner.rb:550]: yes it is. removed in an upcoming commit.

File: test/test-color-scheme.rb (hunk 6 above)
Comment (nevans) [reply test/test-color-scheme.rb:84]: Oh, I was skimming quickly and hadn't noticed it was in a namespaced class. Oops. Will fix.

File: test/test-color-scheme.rb (hunk 7 above)
Comment (nevans) [reply test/test-color-scheme.rb:98]: I was just copy-paste coding and didn't know about that feature! I'll gladly convert to using it. :)

File: test/test-color-scheme.rb (hunk 8 above)
Comment (nevans) [reply test/test-color-scheme.rb:108]: Good point. Deleted.

File: lib/test/unit/color-scheme.rb (hunk 2 above)
Comment (nevans) [reply lib/test/unit/color-scheme.rb:20]: Updated `\A` and `\z` and sorted.  IIRC, when I first started the regexp, I had sorted in alphabetical order, but the trailing `|` vs `)` had messed me up and after that I only sorted everything except the last value. :laughing:

Comment (nevans) [reply lib/test/unit/color-scheme.rb:20]: and actually, the list was already sorted alphabetically, if you ignore the trailing `|`. Should `nsterm|` sort before or after `nsterm-build\d+|`? My brain said yes. ASCII sorting says no. :slightly_smiling_face:

Comment (kou) [reply lib/test/unit/color-scheme.rb]: Oh, sorry.

File: lib/test/unit/color-scheme.rb (hunk 3 above)
Comment (kou) [reply lib/test/unit/color-scheme.rb:159]: Thanks. I understand.
(Sorry for my short response against your long answer...)

File: test/test-color-scheme.rb (hunk 5 above)
Comment (kou) [reply test/test-color-scheme.rb:48]: I see.

**Comment by kou** [comment]
Thanks!

File: lib/test/unit/color-scheme.rb (hunk 3 above)
Comment (nevans) [reply lib/test/unit/color-scheme.rb:159]: That's fine! Sorry for my long answer to your short yes or no question! :laughing:

**Comment by nevans** [comment]
Thank you!
